allennlp/semparse/worlds/atis_world.py

```diff
@@ -0,0 +1,133 @@
+        self.valid_actions: Dict[str, List[str]] = self.init_all_valid_actions()
+        self.grammar_str: str = self.get_grammar_str()
+        self.grammar_with_context = Grammar(self.grammar_str)
+
+    def get_valid_actions(self) -> Dict[str, List[str]]:
+        return self.valid_actions
+
+    def init_all_valid_actions(self) -> Dict[str, List[str]]:
+        """
+        We initialize the world's valid actions with that of the context. This means that the strings
+        and numbers that were valid earlier in the interaction are also valid. We then add new valid strings
+        and numbers from the current utterance.
+        """
+        valid_actions = deepcopy(self.sql_table_context.valid_actions)
+        for string in self.get_strings_from_utterance():
+            if string not in valid_actions['string']:
+                valid_actions['string'].append(format_action('string', string))
+
+        numbers = ['0', '1']
+        for utterance in self.utterances:
+            numbers.extend(get_numbers_from_utterance(utterance))
+            for number in numbers:
+                if number not in valid_actions['number']:
+                    valid_actions['number'].append(format_action('number', number))
+
+        return valid_actions
+
+    def get_grammar_str(self) -> str:
+        """
+        Generate a string that can be used to instantiate a ``Grammar`` object. The string is a sequence of
+        rules that define the grammar.
+        """
+        grammar_str_with_context = self.sql_table_context.grammar_str
+        numbers = [number.split(" -> ")[1].lstrip('["').rstrip('"]') for \
+                   number in sorted(self.valid_actions['number'], reverse=True)]
+        strings = [string .split(" -> ")[1].lstrip('["').rstrip('"]') for \
+                   string in sorted(self.valid_actions['string'], reverse=True)]
+
+        grammar_str_with_context += generate_one_of_string("number", numbers)
+        grammar_str_with_context += generate_one_of_string("string", strings)
+        return grammar_str_with_context
+
+
+    def get_strings_from_utterance(self) -> List[str]:
+        """
+        Based on the current utterance, return a list of valid strings that should be added.
+        """
+        strings: List[str] = []
+        trigger_lists = [CITIES, AIRPORT_CODES,
+                         STATES, STATE_CODES,
+                         FARE_BASIS_CODE, CLASS,
+                         AIRLINE_CODE_LIST, DAY_OF_WEEK,
+                         CITY_CODE_LIST, MEALS,
+                         RESTRICT_CODES]
+        trigger_dicts = [CITY_AIRPORT_CODES,
+                         AIRLINE_CODES,
+                         CITY_CODES,
+                         GROUND_SERVICE,
+                         YES_NO,
+                         MISC_STR]
+
+        trigger_dict = get_trigger_dict(trigger_lists, trigger_dicts)
+
+        for tokenized_utterance in self.tokenized_utterances:
+            if tokenized_utterance:
```

Comment:
I'm checking if the utterance is empty, since I'm using `-1` to get the last token in the utterance.  I pass it an empty utterance in my test for the global actions.

Reply:
Do you only expect to have this usage in the test?  Don't write logic just for the test - better to have the test be slightly more complicated than to make something allowed just for the test.  It feels like this should fail if you pass it an empty utterance, because that isn't something you should be doing.

Also, maybe the test for the global actions should be in `SqlTableContext`?  If it's there, this isn't an issue at all, right?